File: nvflare/edge/device/ios/ExampleProject/NVFlareSDK/Training/ETTrainerExecutor.swift
//
//  ETTrainerExecutor.swift
//  NVFlare iOS SDK
//
//  ExecutorTorch executor - handles both direct usage and config-based creation
//

import Foundation

/// ExecutorTorch implementation of NVFlareExecutor
/// Supports both direct instantiation and config-based creation via ComponentCreator
/// Clean architecture: Swift coordination → Direct C++ ExecutorTorch integration
public class ETTrainerExecutor: NSObject, NVFlareExecutor, ComponentCreator {
    private let trainerArgs: [String: Any]
    
    // MARK: - ComponentCreator (Resolver)
    
    /// Creates ETTrainerExecutor instances from config
    public static func create(name: String, args: [String: Any]) -> Any {
        return ETTrainerExecutor(args: args)
    }
    
    // MARK: - Initializers
    
    /// Default initializer for direct usage
    public override init() {
        self.trainerArgs = [:]
        super.init()
    }
    
    /// Config-based initializer with stored args
    public init(args: [String: Any]) {
        self.trainerArgs = args
        super.init()
    }
    
    // MARK: - NVFlareExecutor
    
    public func execute(taskData: NVFlareDXO, ctx: NVFlareContext, abortSignal: NVFlareSignal) -> NVFlareDXO {
        print("ETTrainerExecutor: Starting execute() with ExecutorTorch")
        
        // Extract model data from DXO
        guard let modelData = taskData.data["model"] as? String else {
            print("ETTrainerExecutor: No model data found in task data")
            return NVFlareDXO(dataKind: "error", data: ["error": "No model data found"])
        }
        
        // Merge stored args with task data meta (task data takes precedence)
        var finalConfig = trainerArgs
        for (key, value) in taskData.meta {
            finalConfig[key] = value
        }
        
        // Get training parameters from merged config
        let epochs = finalConfig["num_epochs"] as? Int ?? 1
        let batchSize = finalConfig["batch_size"] as? Int ?? 32
        let learningRate = finalConfig["learning_rate"] as? Float ?? 0.01
        
        // Get C++ dataset from FlareRunner context (set during FlareRunner initialization)
        guard let cppDatasetPtr = ctx[NVFlareContextKey.dataset] as? UnsafeMutableRawPointer else {
            print("❌ ETTrainerExecutor: No C++ dataset found in context")
            return NVFlareDXO(dataKind: "error", data: ["error": "No C++ dataset found in context"])
        }
        
        print("🔍 ETTrainerExecutor: Retrieved dataset pointer from context: \(cppDatasetPtr)")
        print("📊 ETTrainerExecutor: Training config - epochs: \(epochs), batch: \(batchSize), lr: \(learningRate)")
        if !trainerArgs.isEmpty {
            print("📊 ETTrainerExecutor: Using config args: \(trainerArgs.keys.sorted())")
        }
        print("📊 ETTrainerExecutor: Using app's C++ dataset directly")
        
        // Create training configuration dictionary for C++ layer
        let trainingConfig: [String: Any] = [
            "num_epochs": epochs,
            "batch_size": batchSize,
            "learning_rate": learningRate,
            "method": finalConfig["method"] as? String ?? "cnn",
            "dataset_shuffle": finalConfig["dataset_shuffle"] as? Bool ?? true
        ]
        
        // Direct approach - pass C++ dataset directly to ETTrainer
        print("ETTrainerExecutor: Passing C++ dataset directly to ETTrainer")
        
        // Create ETTrainer with app's C++ dataset
        let trainer = ETTrainer(
            modelBase64: modelData,
            meta: trainingConfig,
            dataset: cppDatasetPtr  // ← App's C++ ETDataset* directly!
        )
        
        // Handle genuine initialization failures (corrupt model, dataset issues, etc.)
        guard let trainer = trainer else {
            print("ETTrainerExecutor: ETTrainer initialization failed - check model data, dataset type, and ExecutorTorch dependencies")
            return NVFlareDXO(dataKind: "error", data: [
                "error": "ETTrainer initialization failed",
                "details": "Possible causes: invalid model data, unsupported dataset type, or ExecutorTorch loading issues",
                "config_args": trainerArgs
            ])
        }
        
        // Execute training (C++ ExecutorTorch does the heavy lifting)
        print("ETTrainerExecutor: Starting ExecutorTorch training...")
        let result = trainer.train()
        
        print("ETTrainerExecutor: Training completed with \(result.keys.count) result keys")
        
        // Convert result to DXO format
        return NVFlareDXO(
            dataKind: "model",
            data: result,
            meta: [
                "training_completed": true,
                "timestamp": Date().timeIntervalSince1970,
                "epochs": epochs,
                "learning_rate": learningRate,
                "batch_size": batchSize,
                "dataset": "app_cpp_dataset",
                "architecture": "executorch_native",
                "config_args": trainerArgs.isEmpty ? "none" : trainerArgs.keys.sorted()
            ]
        )
    }
} 

File: nvflare/edge/device/android/sdk/ETTrainerExecutor.kt
package com.nvidia.nvflare.sdk

import android.util.Log
import com.nvidia.nvflare.sdk.core.DXO
import com.nvidia.nvflare.sdk.core.Context
import com.nvidia.nvflare.sdk.core.ContextKey
import com.nvidia.nvflare.sdk.core.Signal
import com.nvidia.nvflare.sdk.core.Executor
import com.nvidia.nvflare.sdk.core.AndroidFlareRunner
import com.nvidia.nvflare.sdk.training.ETTrainer
import com.nvidia.nvflare.sdk.training.TrainingConfig
import com.nvidia.nvflare.sdk.utils.TaskHeaderKey
import kotlinx.coroutines.runBlocking

/**
 * ExecuTorch-specific implementation of Executor.
 * Handles ET training execution with dataset creation and result conversion.
 */
class ETTrainerExecutor(
    private val trainer: ETTrainer,
    private val storedArgs: Map<String, Any> = emptyMap()
) : Executor {
    private val TAG = "ETTrainerExecutor"

    override fun execute(taskData: DXO, ctx: Context, abortSignal: Signal): DXO {
        if (abortSignal.isTriggered) {
            throw RuntimeException("Execution aborted")
        }

        try {
            Log.d(TAG, "Starting training execution")
            Log.d(TAG, "Task data keys: ${taskData.data.keys}")
            
            // Extract training configuration from task data
            val trainingConfig = extractTrainingConfig(taskData, ctx)
            
            // Extract model data from task data
            val modelData = taskData.data["model"] as? String
            
            // Validate model data
            if (modelData.isNullOrEmpty()) {
                Log.e(TAG, "No model data found in task data")
                throw RuntimeException("No model data found in task data")
            }
            
            // Get dataset from context (iOS pattern)
            val dataset = ctx[ContextKey.DATASET] as? com.nvidia.nvflare.sdk.core.Dataset
                ?: throw RuntimeException("No dataset found in context")
            
            Log.d(TAG, "Retrieved dataset from context: ${dataset.javaClass.simpleName}, size: ${dataset.size()}")
            
            // Set dataset on trainer
            trainer.setDataset(dataset)
            
            // Execute training using ETTrainer directly
            val result = runBlocking {
                trainer.train(trainingConfig, modelData)
            }
            
            Log.d(TAG, "Training completed successfully")
            
            // Convert result back to DXO
            return DXO.fromMap(result)
            
        } catch (e: Exception) {
            Log.e(TAG, "Training execution failed", e)
            throw RuntimeException("Training failed: ${e.message}", e)
        }
    }
    
    private fun extractTrainingConfig(taskData: DXO, ctx: Context): TrainingConfig {
        // Merge stored args with task data meta (task data takes precedence)
        // This matches the iOS pattern where job-level config is merged with per-task config
        val finalConfig = storedArgs.toMutableMap()
        
        // Task data meta overrides stored args
        val taskMeta = taskData.meta as? Map<String, Any> ?: emptyMap()
        for ((key, value) in taskMeta) {
            finalConfig[key] = value
        }
        
        // Get job name from runner context
        val runner = ctx[ContextKey.RUNNER] as? AndroidFlareRunner
        val jobName = runner?.jobName ?: ""
        Log.d(TAG, "Job name: '$jobName'")
        
        // Add job name to config data for method determination
        finalConfig[TaskHeaderKey.JOB_NAME] = jobName
        
        return TrainingConfig.fromMap(finalConfig)
    }
}

/**
 * Factory for creating ETTrainerExecutor instances.
 */
object ETTrainerExecutorFactory {
    
    fun createExecutor(context: android.content.Context, method: String, meta: Map<String, Any>): ETTrainerExecutor {
        // Create trainer with meta (used for artifact management)
        val trainer = ETTrainer(context, meta)
        
        // Pass meta as stored args to executor (matches iOS pattern)
        // These stored args will be merged with per-task config, with task config taking precedence
        return ETTrainerExecutor(trainer, meta)
    }
}
